baseline-error-prone/src/main/java/com/palantir/baseline/errorprone/StringBuilderConstantParameters.java

```diff
@@ -0,0 +1,164 @@
+                    .named("toString")
+                    .withParameters();
+
+    @Override
+    public Description matchMethodInvocation(MethodInvocationTree tree, VisitorState state) {
+        if (!STRING_BUILDER_TO_STRING.matches(tree, state)) {
+            return Description.NO_MATCH;
+        }
+        Optional<List<ExpressionTree>> result = tree.getMethodSelect().accept(StringBuilderVisitor.INSTANCE, state);
+        if (!result.isPresent()) {
+            return Description.NO_MATCH;
+        }
+        List<ExpressionTree> arguments = result.get();
+        Stream<String> prefixStream = arguments.stream().findFirst()
+                .map(ASTHelpers::getType)
+                .filter(type ->
+                        ASTHelpers.isSameType(type, state.getTypeFromString("java.lang.String"), state))
+                .map(ignored -> Stream.<String>of())
+                .orElseGet(() -> Stream.of("\"\""));
+
+        return buildDescription(tree)
+                .setMessage(MESSAGE)
+                .addFix(SuggestedFix.builder()
+                        .replace(tree, Streams.concat(prefixStream, arguments.stream()
+                                .map(node -> getArgumentSourceString(state, node)))
+                                .collect(Collectors.joining(" + ")))
+                        .build())
+                .build();
+    }
+
+    private static String getArgumentSourceString(VisitorState state, ExpressionTree tree) {
+        String originalSource = state.getSourceForNode(tree);
+        // Ternary expressions must be parenthesized to avoid leaking into preceding or following expressions.
+        if (tree instanceof ConditionalExpressionTree || tree instanceof BinaryTree) {
+            return '(' + originalSource + ')';
+        }
+        return originalSource;
+    }
+
+    /**
+     * {@link StringBuilderVisitor} checks if a {@link StringBuilder#toString()} invocation can be followed up
+     * a fluent invocation chain, therefore must have a constant number of arguments.
+     * If so, the visitor results in a present {@link Optional} of {@link ExpressionTree arguments} in the order
+     * they are {@link StringBuilder#append(Object) appended}, otherwise an {@link Optional#empty() empty optional}
+     * is returned.
+     * This allows us to maintain a single implementation for validation and building a {@link SuggestedFix} without
+     * sacrificing build time allocating objects for {@link StringBuilder builders} which don't fit our pattern.
+     */
+    private static final class StringBuilderVisitor
+            extends SimpleTreeVisitor<Optional<List<ExpressionTree>>, VisitorState> {
+        private static final StringBuilderVisitor INSTANCE = new StringBuilderVisitor();
+
+        private StringBuilderVisitor() {
+            super(Optional.empty());
+        }
+
+        @Override
+        public Optional<List<ExpressionTree>> visitNewClass(NewClassTree node, VisitorState state) {
```

Comment:
What is the semantic difference between an absent list and present empty list?

Reply:
Documented here: https://github.com/palantir/gradle-baseline/pull/832/files#diff-515ce9d314ac0796483b739fc5035b83R104-R110

Optional of empty list will result from:
```java
return new StringBuilder().toString();
```

Which should be replaced with:
```java
return "";
```